docker_utils/rtenvsub.sh

```diff
@@ -0,0 +1,434 @@
+            exit 64
+        fi
+
+        mapfile -t directories < <(find "${full_path}" -mindepth 1 -type d -exec readlink -m {} \;)
+        mapfile -t files < <(find "${full_path}" -type f -exec readlink -m {} \;)
+
+        # Create directory structure, check if destination is empty
+        if [ -n "$(ls -A "${destination}")" ] && ! ${overlay} ; then
+            color_echo red "Destination directory is not empty, if you still want to overlay into it please use the -o/--overlay option"
+            print_usage
+            exit 1
+        else
+            for dir in "${directories[@]}"; do
+                create_directory_structure "${destination}" "${dir}" "${full_path}"
+            done
+
+        fi
+
+        # Create named pipes / files and set up cleanup on signals for them
+        if [ -z "${files[*]}" ] ; then
+            color_echo magenta "Destination directory does not contain any files, no pipes created for ${full_path}!"
+        else
+            for file in "${files[@]:-}"; do
+                add_on_sig "rm -f ${destination}/${file#${full_path}}"
+                if ${nofifo} ; then
+                    render_file "${destination}" "${file}" "${full_path}"
+                else
+                    setup_named_pipe "${destination}" "${file}" "${full_path}" &
+                fi
+            done
+        fi
+
+        # Set up safe cleanup for directory structure (needs to be done in
+        # reverse order to ensure safety of operation without recursive rm
+        local index
+        for (( index=${#directories[@]}-1 ; index>=0 ; index-- )) ; do
+            add_on_sig "rmdir ${destination}/${directories[${index}]#${full_path}}"
+        done
+
+        # Run update loop and detach it
+        if ${daemonize} ; then
+            inotify_looper "${destination}" "${full_path}" &
+        else
+            inotify_looper "${destination}" "${full_path}" &
+        fi
+        looper_pids+=( "${!}" )
+    done
+    if ! ${daemonize} ; then
+        debug 8 "Waiting for looper pids: ${looper_pids[*]}"
+        wait "${looper_pids[*]}"
+    fi
+}
+
+# Unit tests
+# shellcheck disable=SC2046,SC2154,SC2016,SC2034,SC2064
+function unit_tests {
+    export verbosity=10
+    debug 5 "Running unit tests!"
+    # Basic setup
+    export TEST_VARIABLE1='/dev/null'
+    export TEST_VARIABLE2='example.com'
+    create_secure_tmp tmp_source_test_dir 'dir'
+    create_secure_tmp tmp_dest_test_dir 'dir'
+    create_secure_tmp tmp_source_test_file 'file' "${tmp_source_test_dir}"
+    test_string=$(tr -dc '[:alnum:]' < /dev/urandom | fold -w 1024 | head -n 1)
+    export signal='SIGUSR1'
+    # Set up a proces to listen to signals and perform actions
+    signal_test_file="${tmp_source_test_dir}/signal_test_file"
+    process="$(signal_processor "${signal}"  "test -f ${signal_test_file} && echo ${test_string} > ${signal_test_file}")"
+    export process
+
+    # Test setting up a named pipe
```

Comment:
Might be worth it to split each test into it's own function.

Reply:
There is quite a bit of setup and teardown that's shared at the moment so it would be a bit more effort and would make the tests take longer. Do you find them to be confusing or just thinking for function/code re-usability?

Reply:
Just for re-usability, if we add more functionality it's nice to just have to write a single function but I see that the setup and teardown stuff would be a bit of work to move out. I'm fine with leaving it as-is, we can address these tests later if we decide on a framework or more general testing method.

Reply:
oks, going to merge as is but we can always refactor later.